### ISSUE TITLE
High contention on Worker thread pool queue

### ISSUE THREAD

**Issue by johnaohara** [eclipse-vertx/vert.x#3184]
In VertxImpl, a 'Executors.newFixedThreadPool' is created for the worker thread pool ;

```
    ExecutorService workerExec = Executors.newFixedThreadPool(options.getWorkerPoolSize(),
        new VertxThreadFactory("vert.x-worker-thread-", checker, true, options.getMaxWorkerExecuteTime(), options.getMaxWorkerExecuteTimeUnit()));
```

Executors.newFixedThreadPool() allocates a LinkedBlockingQueue for the workQueue;

```
    public static ExecutorService newFixedThreadPool(int nThreads, ThreadFactory threadFactory) {
        return new ThreadPoolExecutor(nThreads, nThreads,
                                      0L, TimeUnit.MILLISECONDS,
                                      new LinkedBlockingQueue<Runnable>(),
                                      threadFactory);
    }
```

LinkedBlockingQueue shows up as a scalability bottleneck when heavily contended in systems where there are many cores (32 in our test case), the tasks delegated to the worker pool are fast and do not make blocking calls. We were unable to saturate 32 CPU cores with current implmentation.

Swapping the implementation to a data structure that scales better with high contention, we observed max throughput increase from 182k req/sec to 290k req/sec, and CPU cores became saturated.

**Comment by johnaohara** [comment]
#3182 introduces a configurable WorkerExecutorFactory to allow users to substitute the default implementation with a custom implementation

**Comment by rvansa** [comment]
The problem is probably going to be solved through #3193 .
